python/pyproject.toml

```diff
@@ -41,14 +42,16 @@ classifiers = [
 
 [project.optional-dependencies]
 test = [
+    "cloudpickle",
+    "cudf==23.12.*",
+    "cupy",
+    "dask>=2023.1.1",
+    "distributed>=2023.1.1",
+    "numba>=0.57.1",
     "pytest",
     "pytest-asyncio",
-    "cloudpickle",
-    "cudf",
-    "dask",
-    "distributed",
-    "cupy-cuda11x",
-]
+    "pytest-rerunfailures",
+] # This list was generated by `rapids-dependency-file-generator`. To make changes, edit ../dependencies.yaml and run `rapids-dependency-file-generator`.
```

Comment:
How did `rapids-dependency-file-generator` come up with this list? In particular the Dask/Distributed pins seem very arbitrary.

Reply:
These are coming from `dependencies.yaml`. That file is supposed to be a "single source of truth" for all dependencies and their pinnings (with the exception of meta.yaml conda recipes and CMakeLists.txt which are not implemented at this time). Today, `rapids-dependency-file-generator` generates `pyproject.toml`, developer conda environments, and can generate dependency lists on-the-fly that aren't committed to the repo but are used during various builds (e.g. devcontainers, rapids-compose).

Reply:
I just noticed that `conda/environments/ucxx-cuda118_arch-x86_64.yaml` is not being generated by `dependencies.yaml` so I'll try to adjust that.

Reply:
There are some features like channel pinnings (`dask/label/dev::dask`) and comments that would be lost in the developer environment. @pentschev Do you have thoughts on whether we should rework this to be auto-generated like other RAPIDS packages or would you prefer to maintain it separately?

Reply:
See what you think of commit 4fc2c974dbff425f73aaf711fbf22dad6571d113. I added a `dev` section with the dependencies like `libtool` and `autoconf` and `dask-cuda` that aren't otherwise specified by ucxx as build/run dependencies.

Reply:
> These are coming from `dependencies.yaml`. That file is supposed to be a "single source of truth" for all dependencies and their pinnings (with the exception of meta.yaml conda recipes and CMakeLists.txt which are not implemented at this time). Today, `rapids-dependency-file-generator` generates `pyproject.toml`, developer conda environments, and can generate dependency lists on-the-fly that aren't committed to the repo but are used during various builds (e.g. devcontainers, rapids-compose).

Sorry, this was entirely my mistake. I knew it was coming from `dependencies.yaml` but I didn't realize it was being pinned to `>=2023.1.1` there. I'm guessing when this was last adjusted `python/pyproject.toml` was not regenerated so I got confused by this seemingly arbitrary versioning.

Reply:
> There are some features like channel pinnings (`dask/label/dev::dask`) and comments that would be lost in the developer environment. @pentschev Do you have thoughts on whether we should rework this to be auto-generated like other RAPIDS packages or would you prefer to maintain it separately?

I guess that has been addressed by https://github.com/rapidsai/ucxx/commit/4fc2c974dbff425f73aaf711fbf22dad6571d113, is that right? If so I think this question isn't relevant anymore.

> See what you think of commit [4fc2c97](https://github.com/rapidsai/ucxx/commit/4fc2c974dbff425f73aaf711fbf22dad6571d113). I added a `dev` section with the dependencies like `libtool` and `autoconf` and `dask-cuda` that aren't otherwise specified by ucxx as build/run dependencies.

I see that it included `dask/label/dev` as well which is great and addresses your question from above, these changes are perfect, thanks for doing that Bradley!